### DISCUSSION TITLE
Different database for each user

### DISCUSSION THREAD

**Discussion by afonseca69** [simolus3/drift#1464] [Q&A] [answered]
Hi, In my flutter app I create a database file named for each login user. Each user has his own database. So when I log out and login with another user, I close the database, but it did not open again after the new user login. So is there a way to do this? To open an exist database or create a new one, I use LazyDatabase that returns VmDatabase.

```dart
LazyDatabase _openConnection() {
return LazyDatabase(() async {
final dbFolder = await getApplicationDocumentsDirectory();
//final dbFolder = await getExternalStorageDirectory();
print(dbFolder.path);
final file = File(p.join(dbFolder.path, '$bdinscricao.sqlite'));
print(file);
vgbdaberto = true;
return VmDatabase(file, logStatements: true);
});
}
class DbGestor extends _$DbGestor implements IDbGestor {
static DbGestor instance = DbGestor._internal();
late ClienteDAO clienteDAO;
late ParametroDAO parametroDAO;
late PedidoDAO pedidoDAO;
late PedidoItemDAO pedidoitemDAO;
late PrecoDAO precoDAO;
late ProdutoDAO produtoDAO;
late TituloDAO tituloDAO;

DbGestor._internal() : super(_openConnection()) {
clienteDAO = ClienteDAO(this);
parametroDAO = ParametroDAO(this);
pedidoDAO = PedidoDAO(this);
pedidoitemDAO = PedidoItemDAO(this);
precoDAO = PrecoDAO(this);
produtoDAO = ProdutoDAO(this);
tituloDAO = TituloDAO(this);
}
```

**Comment by simolus3** [comment]
> I close the database, but it did not open again after the new user login

How are you creating new databases? In addition to closing the old instance, you need to create a new instance of your database class and use it to open a new database connection.

**Comment by afonseca69** [comment, marked as answer]
Thank you for the reply.

I have two databases. One for configuration and some internal tables.

The other one, is the user database. This one has it's file name according to the login user.

So, when I log out I should close de database, but when I login the database is still in memory, and did not.

I use, mobx and flutter_modular

I use DAO files for each table.

[cid:d0c74e87-6b7b-481e-9d44-58993f75508d]

dbgestor is the user database

I'm add dbgestor.dart for the code

When I call DbGestor.instance the first time, it creates or open the database, and it continues openned until I close the app.

DbGestor dbgt = DbGestor.instance;

I have tried de dispose command on the login but it is crashing the app.


Thank you very much.





Atenciosamente,

[SoftServ Sistemas]<http://www.softserv.com.br/>[Google+]<https://plus.google.com/+SoftservBr1994> [Curta nossa pÃ¡gina] <https://www.facebook.com/softservsistemas/>  [Conecte-se ao Linkedin] <https://www.linkedin.com/company/softserv-sistemas>  [Siga-nos no Twitter] <https://twitter.com/softserv1994>  [Siga-nos no Instagram] <https://www.instagram.com/softservsistemas>  [Inscreva-se em nosso canal] <http://www.youtube.com/c/SoftservBr1994>    André Fonseca
Pesquisa e Desenvolvimento
Suporte
skype.softserv
(11) 3995 5200
(21) 3195 5200
(31) 3195 5200
(41) 3195 5200
(51) 3195 5200

Comercial
0800 727 2099
Ligação gratuita
São Paulo
Rio de Janeiro
Minas Gerais
Paraná
Rio Grande do Sul


Brasil
[Veja no ***@***.***,-43.271048,17z/data=!4m5!1m2!2m1!1ssoftserv!3m1!1s0x0000000000000000:0xa22666f336c06271?hl=pt-BR>    Sede
Av. Pastor Martin Luther King Jr. 126 Office 2000 Sala 226,
Shopping Nova América, Rio de janeiro - RJ, 20765-971   [http://softserv.com.br/email/assinatura/PENSE_IMPRIMIR.gif]
________________________________



________________________________
De: Simon Binder ***@***.***>
Enviado: sexta-feira, 24 de setembro de 2021 17:08
Para: simolus3/moor ***@***.***>
Cc: Andre Fonseca ***@***.***>; Author ***@***.***>
Assunto: Re: [simolus3/moor] Different database for each user (Discussion #1464)


I close the database, but it did not open again after the new user login

How are you creating new databases? In addition to closing the old instance, you need to create a new instance of your database class and use it to open a new database connection.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/simolus3/moor/discussions/1464#discussioncomment-1381851>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AHLA7IXHLFLDVVARIDHAPT3UDTLFRANCNFSM5EV3BSWA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

**Reply by simolus3** [reply]
After you close the `DbGestor` instace, you should not use it anymore. So you'll probably want something like this:

```dart
DbGestor.instance.close();
DbGestor.instance = new DbGestor();
```

**Comment by afonseca69** [comment]
Thank you very much.

It is ok now


Atenciosamente,

[SoftServ Sistemas]<http://www.softserv.com.br/>[Google+]<https://plus.google.com/+SoftservBr1994> [Curta nossa pÃ¡gina] <https://www.facebook.com/softservsistemas/>  [Conecte-se ao Linkedin] <https://www.linkedin.com/company/softserv-sistemas>  [Siga-nos no Twitter] <https://twitter.com/softserv1994>  [Siga-nos no Instagram] <https://www.instagram.com/softservsistemas>  [Inscreva-se em nosso canal] <http://www.youtube.com/c/SoftservBr1994>    André Fonseca
Pesquisa e Desenvolvimento
Suporte
skype.softserv
(11) 3995 5200
(21) 3195 5200
(31) 3195 5200
(41) 3195 5200
(51) 3195 5200

Comercial
0800 727 2099
Ligação gratuita
São Paulo
Rio de Janeiro
Minas Gerais
Paraná
Rio Grande do Sul


Brasil
[Veja no ***@***.***,-43.271048,17z/data=!4m5!1m2!2m1!1ssoftserv!3m1!1s0x0000000000000000:0xa22666f336c06271?hl=pt-BR>    Sede
Av. Pastor Martin Luther King Jr. 126 Office 2000 Sala 226,
Shopping Nova América, Rio de janeiro - RJ, 20765-971   [http://softserv.com.br/email/assinatura/PENSE_IMPRIMIR.gif]
________________________________



________________________________
De: Simon Binder ***@***.***>
Enviado: segunda-feira, 27 de setembro de 2021 10:03
Para: simolus3/moor ***@***.***>
Cc: Andre Fonseca ***@***.***>; Author ***@***.***>
Assunto: Re: [simolus3/moor] Different database for each user (Discussion #1464)


After you close the DbGestor instace, you should not use it anymore. So you'll probably want something like this:

DbGestor.instance.close();
DbGestor.instance = new DbGestor();

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/simolus3/moor/discussions/1464#discussioncomment-1390248>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AHLA7IXX2ES36G63RIVP3KTUEBTRZANCNFSM5EV3BSWA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
